### PR TITLE
[6.3] Provision a host with RHEV CR and check the hypervisor VM OS type (BZ:1315281)

### PR DESCRIPTION
**Pull request by ldjebran**
cover https://bugzilla.redhat.com/show_bug.cgi?id=1315281
depend on : https://github.com/SatelliteQE/nailgun/pull/469

```bash
pytest -v tests/foreman/ui/test_computeresource_rhev.py::RhevComputeResourceHostTestCase::test_positive_check_provisioned_rhev_os
================================================= test session starts ==================================================
platform linux2 -- Python 2.7.13, pytest-3.2.3, py-1.5.2, pluggy-0.4.0 -- /home/dlezz/.pyenv/versions/2.7.13/envs/sat-6.3.0/bin/python2.7
cachedir: .cache
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: xdist-1.20.1, services-1.2.1, mock-1.6.3, forked-0.2, cov-2.5.1, fauxfactory-1.0.1
collected 1 item                                                                                                        
2017-12-05 12:22:22 - conftest - DEBUG - BZ deselect is disabled in settings


tests/foreman/ui/test_computeresource_rhev.py::RhevComputeResourceHostTestCase::test_positive_check_provisioned_rhev_os <- robottelo/decorators/__init__.py PASSED

============================================= 1 passed in 1077.65 seconds ==============================================
```
The os type keys has been updated by this PR:
https://gerrit.ovirt.org/#/c/26569/3/packaging/conf/osinfo-defaults.properties

### CODE REVIEW

File: robottelo/ui/computeresource.py [hunk 1]
```diff
@@ -191,6 +195,17 @@ def set_values(self, **kwargs):
...
             self.set_value(template_key, template)
             del kwargs[template_key]
+        memory = kwargs.get('memory')
+        if memory is not None:
+            # to make the memory value take effect we have to set it's hidden
+            # input value
+            memory_hidden_input = self.page.wait_until_element_exists(
+                self.memory_locator_hidden)
+            # memory must be in bytes, 1 MB = 1024*1024 = 1048576 bytes
+            self.page.browser.execute_script(
```
Comment (oshtaier) [inline robottelo/ui/computeresource.py]: sorry, but can you provide screenshot?
Are you trying to perform action that real user cannot do?

Comment (ldjebran) [reply robottelo/ui/computeresource.py]: Seems we also cannot do what the user can do as, seems there is a script that handle that field and copy the value to the hidden field , eg: when send keys by selenium to memory field  the hidden field is not updated and the value is not updated.
(of course we can see the  new value in memory field but that value is not saved as not detected )
the only way that I found is to explicitly set that hidden filed value.

Comment (oshtaier) [reply robottelo/ui/computeresource.py]: hmm, I think it should be investigated more as `send_keys` does exactly the same as when user input data from keyboard. Please, look what events are called and what trigger copy paste from real field to hidden.
Sorry, but proposed solution is not acceptable from WebUI test automation perspective

Comment (ldjebran) [reply robottelo/ui/computeresource.py]: @oshtaier think I found how to explicitly fire that event, without setting the hidden value, will make the necessary changes

**Comment by ldjebran** [comment]
@oshtaier updated with triggering the change event + add more ajax time , as many tasks happen in the background and the test randomly failed on "Timeout waiting for page to load" on submit.
```
pytest -v tests/foreman/ui/test_computeresource_rhev.py::RhevComputeResourceHostTestCase::test_positive_check_provisioned_rhev_os
================================================= test session starts ==================================================
platform linux2 -- Python 2.7.13, pytest-3.2.3, py-1.5.2, pluggy-0.4.0 -- /home/dlezz/.pyenv/versions/2.7.13/envs/sat-6.3.0/bin/python2.7
cachedir: .cache
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: xdist-1.20.1, services-1.2.1, mock-1.6.3, forked-0.2, cov-2.5.1, fauxfactory-1.0.1
collected 1 item                                                                                                        
2017-12-05 16:05:31 - conftest - DEBUG - BZ deselect is disabled in settings


tests/foreman/ui/test_computeresource_rhev.py::RhevComputeResourceHostTestCase::test_positive_check_provisioned_rhev_os <- robottelo/decorators/__init__.py PASSED

============================================= 1 passed in 1645.91 seconds ==============================================
```

File: robottelo/ui/computeresource.py [hunk 2]
```diff
@@ -63,6 +63,7 @@ def _assign_locator_value(self, target, value):
             pass
         else:
             self.page.assign_value(target, value)
+        self.page.wait_for_ajax()
```
Comment (oshtaier) [inline robottelo/ui/computeresource.py]: both `click` and `assign_value` do `wait_for_ajax` in the end of their execution, so why do we need it?

File: robottelo/ui/computeresource.py [hunk 3]
```diff
@@ -191,6 +193,20 @@ def set_values(self, **kwargs):
         if template is not None:
             self.set_value(template_key, template)
             del kwargs[template_key]
+        # when setting memory value it does not fire the change event,
+        # that do the necessary validation and update the memory hidden field,
+        # without this event fired the memory value cannot be saved,
+        memory_key = 'memory'
```
Comment (oshtaier) [inline robottelo/ui/computeresource.py]: don't like that `set_value` method at all as too much of redundant code, other than that, looks good

File: tests/foreman/ui/test_computeresource_rhev.py [hunk 4]
```diff
@@ -1131,3 +1135,168 @@ def test_positive_provision_rhev_with_host_group(self):
...
+        :expectedresults: Hypervisor VM OS type must be the same as the one
+            used in provisioning.
+
+        :CaseAutomation: Automated
+
+        :BZ: 1315281
+
+        :CaseLevel: System
+        """
+        host_name = gen_string('alpha', 9).lower()
```
Comment (oshtaier) [inline tests/foreman/ui/test_computeresource_rhev.py]: any necessity to specify all these strings length?

File: tests/foreman/ui/test_computeresource_rhev.py [hunk 5]
```diff
@@ -1131,3 +1135,168 @@ def test_positive_provision_rhev_with_host_group(self):
...
+            # redirected to the created host
+            # wait until redirected to host page
+            self.assertIsNotNone(
+                session.hosts.wait_until_element(
+                    locators["host.host_page_title"] % vm_host_name,
+                    timeout=60
+                )
+            )
+            # Query RHEV hypervisor for vm description
+            vm_host_data = entities.Host().search_json(
```
Comment (oshtaier) [inline tests/foreman/ui/test_computeresource_rhev.py]: any necessity to use direct `search_json`. Such approach is strongly discouraged

File: tests/foreman/ui/test_computeresource_rhev.py [hunk 6]
```diff
@@ -1131,3 +1135,168 @@ def test_positive_provision_rhev_with_host_group(self):
...
+                query=dict(search='name="{0}"'.format(vm_host_name))
+            )['results'][0]
+            # Using documented RHEV API access
+            response = client.get(
```
Comment (oshtaier) [inline tests/foreman/ui/test_computeresource_rhev.py]: Any alternatives to `client.get` here? Such approach is strongly discouraged

File: tests/foreman/ui/test_computeresource_rhev.py (hunk 5 above)
Comment (ldjebran) [reply tests/foreman/ui/test_computeresource_rhev.py]: we need the uuid field that is returned only by read_json, read_json is a public entity function

File: tests/foreman/ui/test_computeresource_rhev.py (hunk 6 above)
Comment (ldjebran) [reply tests/foreman/ui/test_computeresource_rhev.py]: there is an expensive alternative, is to install a third party library RHEV sdk, also there is some libraries that do the same, but do not see the plus to do so for a one line command.
do not agree with you that this is discouraged as we are using a documented RHEV API call, using a standard http client by the way the same client used by nailgun, please let me know what is discouraged here?

File: robottelo/ui/computeresource.py (hunk 2 above)
Comment (ldjebran) [reply robottelo/ui/computeresource.py]: assign_value do not do wait_for_ajax, and we need that to be performed

File: tests/foreman/ui/test_computeresource_rhev.py (hunk 4 above)
Comment (ldjebran) [reply tests/foreman/ui/test_computeresource_rhev.py]: No, but I can change that, if you insist

File: robottelo/ui/computeresource.py (hunk 2 above)
Comment (ldjebran) [reply robottelo/ui/computeresource.py]: sorry you right, the select and input instead does do  wait_for_ajax

File: tests/foreman/ui/test_computeresource_rhev.py (hunk 6 above)
Comment (ldjebran) [reply tests/foreman/ui/test_computeresource_rhev.py]: Note client.get  == requests.get , if you want I can invoke requests.get directly, but I do so I have to add it to requirements.

Comment (oshtaier) [reply tests/foreman/ui/test_computeresource_rhev.py]: nope, that's fine. I meant client.get is worse solution in comparison with `nailgun.entity.read()`
Check that out:
https://github.com/SatelliteQE/robottelo/issues/2837
So you are adding the code that should be refactored in future in advance. That is not good idea, but if no way and you sure on 100% that you cannot use nailgun for you purpose, so I will be fine with `client.get`

File: tests/foreman/ui/test_computeresource_rhev.py (hunk 5 above)
Comment (oshtaier) [reply tests/foreman/ui/test_computeresource_rhev.py]: do we have bug for that behavior?

File: tests/foreman/ui/test_computeresource_rhev.py (hunk 4 above)
Comment (oshtaier) [reply tests/foreman/ui/test_computeresource_rhev.py]: only if you are going to have any other modification in the code

File: tests/foreman/ui/test_computeresource_rhev.py (hunk 5 above)
Comment (abalakh) [reply tests/foreman/ui/test_computeresource_rhev.py]: You should probably add `uuid` field instead. Pls refer to https://github.com/SatelliteQE/robottelo/pull/4928#discussion_r126986809 to remind read/read_json difference.

File: tests/foreman/ui/test_computeresource_rhev.py (hunk 6 above)
Comment (abalakh) [reply tests/foreman/ui/test_computeresource_rhev.py]: It's strongly discouraged in most cases for Satellite API (you should use nailgun instead), but that's not the case here. This call is ok unless there's our framework to work with rhev api which i'm not aware of 🙂

**Comment by ldjebran** [comment]
@abalakh @oshtaier all comments addressed
```
(sat-6.3) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v tests/foreman/ui/test_computeresource_rhev.py::RhevComputeResourceHostTestCase::test_positive_check_provisioned_rhev_os
============================================ test session starts =============================================
platform linux2 -- Python 2.7.14, pytest-3.2.3, py-1.5.2, pluggy-0.4.0 -- /home/dlezz/.pyenv/versions/2.7.14/envs/sat-6.3/bin/python2.7
cachedir: .cache
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: xdist-1.20.1, services-1.2.1, mock-1.6.3, forked-0.2, cov-2.5.1, fauxfactory-1.0.1
collected 1 item                                                                                              
2017-12-12 12:53:26 - conftest - DEBUG - BZ deselect is disabled in settings


tests/foreman/ui/test_computeresource_rhev.py::RhevComputeResourceHostTestCase::test_positive_check_provisioned_rhev_os <- robottelo/decorators/__init__.py PASSED

======================================== 1 passed in 1419.96 seconds =========================================
```

File: tests/foreman/ui/test_computeresource_rhev.py [hunk 7]
```diff
@@ -1131,3 +1135,168 @@ def test_positive_provision_rhev_with_host_group(self):
...
+                loc=self.loc_name,
+                parameters_list=host_parameters_list,
+                provisioning_method='image',
+            )
+            vm_host_name = '{0}.{1}'.format(
+                host_name.lower(), domain.name)
+            # the provisioning take some time to finish, when done will be
+            # redirected to the created host
+            # wait until redirected to host page
+            self.assertIsNotNone(
```
Comment (oshtaier) [inline tests/foreman/ui/test_computeresource_rhev.py]: can we check that 
`Status = OK` not `Pending installation` or something?

Comment (oshtaier) [reply tests/foreman/ui/test_computeresource_rhev.py]: to make sure that provisioning actually happened

**Comment by ldjebran** [comment]
@oshtaier if the provisioning does not happen we could not find it the list of hypervisor's vms (that is affected by the last api call) any way will make that assertion check.

**Comment by oshtaier** [comment]
not sure, it can be in a list as a dead vm which was never started

**Comment by ldjebran** [comment]
Added check for host provisioning status and build status + modified the os search
@oshtaier @abalakh comments addressed
```console
(sat-6.3) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v tests/foreman/ui/test_computeresource_rhev.py::RhevComputeResourceHostTestCase::test_positive_check_provisioned_rhev_os
============================================ test session starts =============================================
platform linux2 -- Python 2.7.14, pytest-3.2.3, py-1.5.2, pluggy-0.4.0 -- /home/dlezz/.pyenv/versions/2.7.14/envs/sat-6.3/bin/python2.7
cachedir: .cache
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: xdist-1.20.1, services-1.2.1, mock-1.6.3, forked-0.2, cov-2.5.1, fauxfactory-1.0.1
collected 1 item                                                                                              
2017-12-12 19:41:22 - conftest - DEBUG - BZ deselect is disabled in settings


tests/foreman/ui/test_computeresource_rhev.py::RhevComputeResourceHostTestCase::test_positive_check_provisioned_rhev_os <- robottelo/decorators/__init__.py PASSED

======================================== 1 passed in 1362.43 seconds =========================================
```

**Pull request merged**